magicgui/_magicgui.py

```diff
@@ -144,36 +187,176 @@ def magicgui(
+        """Return string repr."""
+        params = inspect.signature(magicgui).parameters
+        args = [
+            f"{k}={v!r}"
+            for (k, v) in self.keywords.items()
+            if v not in (params[k].default, {})
+        ]
+        return f"MagicFactory({', '.join(args)})"
+
+    def __call__(self, *args, **kwargs) -> _T:
+        """Call the wrapped _magicgui and return a FunctionGui."""
+        if args:
+            raise ValueError("MagicFactory instance only accept keyword arguments")
+        params = inspect.signature(magicgui).parameters
+        prm_options = self.keywords.pop("param_options", {})
+        prm_options.update({k: kwargs.pop(k) for k in list(kwargs) if k not in params})
+        return self.func(param_options=prm_options, **{**self.keywords, **kwargs})
+
+    def __getattr__(self, name) -> Any:
+        """Allow accessing FunctionGui attributes without mypy error."""
+        pass  # pragma: no cover
+
+    @property
+    def __name__(self) -> str:
+        """Pass function name."""
+        return getattr(self.keywords.get("function"), "__name__", "FunctionGui")
+
+
+# Overloads for magic_factory decorator.  See implementation below
+# TODO: figure out how to get these in a stub file.
+# My first attempts to put them in ``_magicgui.pyi`` broke my type hints in VSCode
+# fmt: off
+@overload  # noqa: E302
+def magic_factory(  # noqa
+    function: Callable[..., _R],
+    *,
+    layout: str = "horizontal",
+    labels: bool = True,
+    tooltips: bool = True,
+    call_button: Union[bool, str] = False,
+    auto_call: bool = False,
+    result_widget: bool = False,
+    main_window: Literal[False] = False,
+    app: AppRef = None,
+    **param_options: dict,
+) -> MagicFactory[FunctionGui[_R]]: ...
+@overload  # noqa: E302
+def magic_factory(  # noqa
+    function: Literal[None] = None,
+    *,
+    layout: str = "horizontal",
+    labels: bool = True,
+    tooltips: bool = True,
+    call_button: Union[bool, str] = False,
+    auto_call: bool = False,
+    result_widget: bool = False,
+    main_window: Literal[False] = False,
+    app: AppRef = None,
+    **param_options: dict,
+) -> Callable[[Callable[..., _R]], MagicFactory[FunctionGui[_R]]]: ...
+@overload  # noqa: E302
+def magic_factory(  # noqa
+    function: Callable[..., _R],
+    *,
+    layout: str = "horizontal",
+    labels: bool = True,
+    tooltips: bool = True,
+    call_button: Union[bool, str] = False,
+    auto_call: bool = False,
+    result_widget: bool = False,
+    main_window: Literal[True],
+    app: AppRef = None,
+    **param_options: dict,
+) -> MagicFactory[MainFunctionGui[_R]]: ...
+@overload  # noqa: E302
+def magic_factory(  # noqa
+    function: Literal[None] = None,
+    *,
+    layout: str = "horizontal",
+    labels: bool = True,
+    tooltips: bool = True,
+    call_button: Union[bool, str] = False,
+    auto_call: bool = False,
+    result_widget: bool = False,
+    main_window: Literal[True],
```

Comment:
Q for my own education: does Literal work if it's not actually a literal? ie will this code pass type checking?

```python
main_win = True
f = magic_factory(func, main_window=main_win)()
f.<something_main_function_gui_provides>()
```
?

Reply:
Or even worse, something where `main_win` is determined from external input...

Reply:
yep, that works

Reply:
Wowza.